asim: fix skew distribution

wenyihu6 opened the pull request:
**asim: fix skew distribution**

Previously, skewedDistribution did not correctly compute and return the ratio
based distribution. This commit fixes the logic to return the ratio.

Epic: none
Release note: none

---

**asim: add TestEvenDistribution**


---

**asim: rename arguments**

This commit renames parameters for distribution helper functions
in pkg/kv/kvserver/asim/state/new_state.go.

---

**asim/state: use epsilon in cumulative weights checking**

Previously, we asserted that cumulative weights in the weighted random
distribution must sum to exactly 1. Due to floating point precision limitations,
this is not always achievable and can lead to false failures like total
cumulative weights for all stores should sum up to 1 but got 1.00. This commit
relaxes the check to allow a small epsilon tolerance in the comparison to avoid
such errors.

---

**asim: improve skewedDistribution and its testing**

Previously, we fixed the skewedDistribution function by generating weights that
decrease by a factor of 1/2 for each subsequent store, and then normalizing them
so that they sum up to 1. The result represented a skewed replica weight
distribution across stores. However, this required two passes: one to generate
and sum the weights, and another to normalize them. This commit improves the
logic by using the finite sum of a geometric series to pre-compute the total,
allowing normalization in a single pass. It also improves test coverage by
adding an echotest that asserts the expected output of the helper functions.

wenyihu6 committed:
asim: fix skew distribution

Previously, skewedDistribution did not correctly compute and return the ratio
based distribution. This commit fixes the logic to return the ratio.

Epic: none
Release note: none

wenyihu6 committed:
asim: add TestEvenDistribution

tbg reviewed (changes requested):
Good catch! Comments aren't major.

It's a shame to churn all of these pngs that I'm about to remove. Maybe you could cherry-pick https://github.com/cockroachdb/cockroach/pull/150847/commits/702aa8b3102aae42e53cd3fe342a82c57912f297 to the front of this commit and regenerate (I'm ok with rewriting history for that :)) ? That should avoid most png churn.

> Reviewed 59 of 59 files at r1, 1 of 1 files at r2, 57 of 57 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151055#-:-OWX7Wuw9dBJ9FmgfeM2:bei0qqh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/asim/state/state_test.go` line 426 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/151055#-OWSsDDq06hdZTueH5Fc:-OWX6ug9Ahs_Dl8sjg6P:b48skat) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b1104c3aae8dae1ef639d097f378a96251dbbec/pkg/kv/kvserver/asim/state/state_test.go#L426)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

Without this PR, the test assigns 100 to stores 1, 2, and 3, and 0 to all others. I might be missing something since I recall during your demo, stores 4, 5, and 6 had non-zero replicas, which I can't explain yet. When I switch to 30 ranges and 10 stores, I get 30 on stores 1–3 and 0 on the rest.
</blockquote></details>

I think the test also had some evenly distributed replicas, I think.
___
*[`pkg/kv/kvserver/asim/state/state_test.go` line 441 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151055#-OWX7LHd-1xK4NEejmnn:-OWX7LHd-1xK4NEejmno:ba4und6) ([raw file](https://github.com/cockroachdb/cockroach/blob/54c6e20ca9e46c6e80322ff97709f8b0330a88d8/pkg/kv/kvserver/asim/state/state_test.go#L441)):*
> ```Go
> 	require.Equal(t, 6, len(stores))
> }
> func TestEvenDistribution(t *testing.T) {
> ```

I'd be fine with testing them all in the same echotest once you have one. I think it will be a good fit.
___
*[`pkg/kv/kvserver/asim/state/new_state.go` line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151055#-OWVpnjy1DoPITmGHjwj:-OWVpnjy1DoPITmGHjwk:bqgj8i6) ([raw file](https://github.com/cockroachdb/cockroach/blob/97bbafe23d131f9104146b3238726c64d7e447bd/pkg/kv/kvserver/asim/state/new_state.go#L42)):*
> ```Go
> }
> 
> func skewedDistribution(n, k int) []float64 {
> ```

`k` (the number of ranges) isn't used. Reading the code, that makes sense to me - the returned slice is used as weights (sum up to one), which can be seen here: 
https://github.com/cockroachdb/cockroach/blob/7b1104c3aae8dae1ef639d097f378a96251dbbec/pkg/kv/kvserver/asim/state/new_state.go#L174-L175

So I'd drop `k` and rename `n` to `stores` to lower confusion. I'd do the same to the other distributions as applicable.
___
*[`pkg/kv/kvserver/asim/state/new_state.go` line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151055#-OWVrHPP9EnQqHVEpPi0:-OWVrHPQBckI30vVfvyf:b-ca20hd) ([raw file](https://github.com/cockroachdb/cockroach/blob/97bbafe23d131f9104146b3238726c64d7e447bd/pkg/kv/kvserver/asim/state/new_state.go#L42)):*
> ```Go
> }
> 
> func skewedDistribution(n, k int) []float64 {
> ```

Could you add a unit test via `echotest`? This could replace your existing test. echotest is great for things like this! The test could print the resulting numbers and their sum for some choices of the store count, so we can see that everything checks out.

This could be

```go
sum := 2.0 - math.Pow(2, float64(-stores))
cur := 1 // then 0.5, 0.25, ...
for i := 0; i < n; i++ {
  weights[i] = cur / sum
  cur /= 2
}
```

This is because

```
\sum_0^{k} 2^{-i} &= \sum_0^\infty 2^{-i} - \sum_{k+1}^\infty 2^-i = 2 - \sum_{k+1}^\infty 2^{-i} = 2 - 2^{-k} \sum_{1}^\infty 2^{-i} = 2 - 2^{-k}
```

![image.png](https://files.reviewableusercontent.io/images/reviewable/5076964/ohD_v66l3jNAd9dTcqoaAdTn/image.png)


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
asim: rename arguments

This commit renames parameters for distribution helper functions
in pkg/kv/kvserver/asim/state/new_state.go.

wenyihu6 committed:
asim: improve skewedDistribution and its testing

Previously, we fixed the skewedDistribution function by generating weights that
decrease by a factor of 1/2 for each subsequent store, and then normalizing them
so that they sum up to 1. The result represented a skewed replica weight
distribution across stores. However, this required two passes: one to generate
and sum the weights, and another to normalize them. This commit improves the
logic by using the finite sum of a geometric series to pre-compute the total,
allowing normalization in a single pass. It also improves test coverage by
adding an echotest that asserts the expected output of the helper functions.

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151055)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/kv/kvserver/asim/state/new_state.go` line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151055#-OWVpnjy1DoPITmGHjwj:-OWYS3UQ1Zp002kpKa8j:b-fznogt) ([raw file](https://github.com/cockroachdb/cockroach/blob/97bbafe23d131f9104146b3238726c64d7e447bd/pkg/kv/kvserver/asim/state/new_state.go#L42)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

`k` (the number of ranges) isn't used. Reading the code, that makes sense to me - the returned slice is used as weights (sum up to one), which can be seen here: 
https://github.com/cockroachdb/cockroach/blob/7b1104c3aae8dae1ef639d097f378a96251dbbec/pkg/kv/kvserver/asim/state/new_state.go#L174-L175

So I'd drop `k` and rename `n` to `stores` to lower confusion. I'd do the same to the other distributions as applicable.
</blockquote></details>

Good idea, done.
___
*[`pkg/kv/kvserver/asim/state/new_state.go` line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151055#-OWVrHPP9EnQqHVEpPi0:-OWYS4Y767imBP_lym_e:bdlkknp) ([raw file](https://github.com/cockroachdb/cockroach/blob/97bbafe23d131f9104146b3238726c64d7e447bd/pkg/kv/kvserver/asim/state/new_state.go#L42)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Could you add a unit test via `echotest`? This could replace your existing test. echotest is great for things like this! The test could print the resulting numbers and their sum for some choices of the store count, so we can see that everything checks out.

This could be

```go
sum := 2.0 - math.Pow(2, float64(-stores))
cur := 1 // then 0.5, 0.25, ...
for i := 0; i < n; i++ {
  weights[i] = cur / sum
  cur /= 2
}
```

This is because

```
\sum_0^{k} 2^{-i} &= \sum_0^\infty 2^{-i} - \sum_{k+1}^\infty 2^-i = 2 - \sum_{k+1}^\infty 2^{-i} = 2 - 2^{-k} \sum_{1}^\infty 2^{-i} = 2 - 2^{-k}
```

![image.png](https://files.reviewableusercontent.io/images/reviewable/5076964/ohD_v66l3jNAd9dTcqoaAdTn/image.png)
</blockquote></details>

:sunglasses: cool trick! (should it be `sum := 2.0 - math.Pow(2, float64(-(stores - 1)))` since we would be going from i = 0 to 2 three stores?)
___
*[`pkg/kv/kvserver/asim/state/state_test.go` line 441 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/151055#-OWX7LHd-1xK4NEejmnn:-OWYT5bx1bv-anXtYbGQ:b-fmxhfn) ([raw file](https://github.com/cockroachdb/cockroach/blob/54c6e20ca9e46c6e80322ff97709f8b0330a88d8/pkg/kv/kvserver/asim/state/state_test.go#L441)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I'd be fine with testing them all in the same echotest once you have one. I think it will be a good fit.
</blockquote></details>

Done. TIL about echotest - neat.
___
*[`pkg/kv/kvserver/asim/state/state_test.go` line 426 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/151055#-OWSsDDq06hdZTueH5Fc:-OWYSsEzCegTuXiUKjHI:bxrx5sa) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b1104c3aae8dae1ef639d097f378a96251dbbec/pkg/kv/kvserver/asim/state/state_test.go#L426)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I think the test also had some evenly distributed replicas, I think.
</blockquote></details>

Ahh, you are right. Makes sense now. 


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
asim/state: use epsilon in cumulative weights checking

Previously, we asserted that cumulative weights in the weighted random
distribution must sum to exactly 1. Due to floating point precision limitations,
this is not always achievable and can lead to false failures like total
cumulative weights for all stores should sum up to 1 but got 1.00. This commit
relaxes the check to allow a small epsilon tolerance in the comparison to avoid
such errors.

wenyihu6 commented:
Last push was to rebase on master to resolve merge conflicts.
